File: windows/platform/file_drop_session.cpp
#include "file_drop_session.h"
#include <flutter/method_result_functions.h>
#include <utility>

namespace share_hub {
using Value = flutter::EncodableValue;
using Map = flutter::EncodableMap;
using List = flutter::EncodableList;
struct FileDropSession::State {
  SourceBridge* source;
  Offer offer;
  Locate locate;
  std::function<void()> error;
  bool listening = false, busy = false, closed = false;
  uint64_t generation = 0;
  List pending;

  void Release(const List& files) {
    if (!source) return;
    for (const auto& file : files) {
      const auto& token = std::get<Map>(file).at(Value("token"));
      std::unique_ptr<SourceBridge::Result> ignore =
          std::make_unique<flutter::MethodResultFunctions<Value>>(nullptr, nullptr, nullptr);
      source->Handle(flutter::MethodCall<Value>("files.release", std::make_unique<Value>(token)), ignore);
    }
  }
  void Finish(bool accepted) {
    if (!accepted) Release(pending);
    pending.clear();
    busy = false;
  }
};
FileDropSession::FileDropSession(SourceBridge& source, Locate locate, Offer offer, std::function<void()> error)
    : state_(std::make_shared<State>()) {
  state_->source = &source; state_->locate = std::move(locate);
  state_->offer = std::move(offer); state_->error = std::move(error);
}
FileDropSession::~FileDropSession() { Close(); }
void FileDropSession::Listen() { if (!state_->closed) state_->listening = true; }
void FileDropSession::Cancel() { state_->listening = false; ++state_->generation; }
bool FileDropSession::ready() const { return !state_->closed && state_->listening && !state_->busy; }
bool FileDropSession::Accept(std::vector<std::wstring> paths, double x, double y) {
  if (!ready()) return false;
  state_->busy = true;
  const auto generation = state_->generation;
  std::weak_ptr<State> weak = state_;
  auto located = std::make_shared<bool>(false);
  state_->locate(x, y, [weak, generation, x, y, located, paths = std::move(paths)](bool accepted) mutable {
    if (*located) return;
    *located = true;
    const auto owner = weak.lock();
    if (!owner || owner->closed) return;
    if (!accepted || !owner->listening || owner->generation != generation) {
      owner->busy = false; return;
    }
    owner->source->AcceptPickedFiles(std::move(paths),
      std::make_unique<flutter::MethodResultFunctions<Value>>(
          [weak, generation, x, y](const Value* value) {
            const auto state = weak.lock();
            if (!state || state->closed) return; // Source store closes on teardown.
            state->pending = std::get<List>(*value);
            if (!state->listening || generation != state->generation) { state->Finish(false); return; }
            auto decided = std::make_shared<bool>(false);
            state->offer(Value(Map{{Value("x"), Value(x)}, {Value("y"), Value(y)},
                                   {Value("files"), Value(state->pending)}}),
                [weak, decided](bool accepted) {
                  if (*decided) return;
                  *decided = true;
                  if (const auto current = weak.lock(); current && !current->closed)
                    current->Finish(accepted);
                });
          },
          [weak, generation](const std::string&, const std::string&, const Value*) {
            if (const auto state = weak.lock(); state && !state->closed) {
              state->busy = false;
              if (state->listening && generation == state->generation) state->error();
            }
          }, nullptr));
  });
  return true;
}
void FileDropSession::Close() {
  if (state_->closed) return;
  Cancel();
  state_->Finish(false);
  state_->closed = true;
  state_->source = nullptr;
  state_->offer = nullptr;
  state_->locate = nullptr;
  state_->error = nullptr;
}
}

File: macos/Platform/Sources/ShareHubPlatform/FileDropSession.swift
import Foundation

/// Main-thread ownership transaction for one OS-originated file batch.
/// Paths never cross the Flutter boundary. Source selection remains asynchronous.
final class FileDropSession {
    typealias Decision = (Bool) -> Void
    typealias Locate = (Double, Double, @escaping Decision) -> Void
    typealias Prepare = ([URL], @escaping (Result<[SelectedFileInfo], Error>) -> Void) -> Void
    typealias Offer = ([SelectedFileInfo], Double, Double, @escaping Decision) -> Void

    private let locate: Locate
    private let prepare: Prepare
    private let release: ([SelectedFileInfo]) -> Void
    private let offer: Offer
    private let onError: () -> Void
    private let holdURLs: ([URL]) -> () -> Void
    private var listening = false, busy = false, closed = false
    private var generation = UUID()
    private var pending: [SelectedFileInfo] = []
    private var lease: URLLease?

    init(locate: @escaping Locate, prepare: @escaping Prepare,
         release: @escaping ([SelectedFileInfo]) -> Void, offer: @escaping Offer,
         onError: @escaping () -> Void,
         holdURLs: @escaping ([URL]) -> () -> Void = { urls in
             let accessed = urls.filter { $0.startAccessingSecurityScopedResource() }
             return { for url in accessed { url.stopAccessingSecurityScopedResource() } }
         }) {
        self.locate = locate; self.prepare = prepare; self.release = release
        self.offer = offer; self.onError = onError; self.holdURLs = holdURLs
    }

    var ready: Bool { precondition(Thread.isMainThread); return listening && !busy && !closed }
    func listen() { precondition(Thread.isMainThread); if !closed { listening = true } }
    func cancel() {
        precondition(Thread.isMainThread)
        listening = false; generation = UUID()
        lease?.stop(); lease = nil
    }

    @discardableResult
    func accept(_ urls: [URL], x: Double, y: Double) -> Bool {
        precondition(Thread.isMainThread)
        guard ready, !urls.isEmpty, urls.count <= SelectedFileStore.maximumFiles,
              urls.allSatisfy({ $0.isFileURL }), x.isFinite, y.isFinite else { return false }
        busy = true
        let generation = self.generation
        let lease = URLLease(holdURLs(urls))
        self.lease = lease
        var located = false
        locate(x, y) { [weak self] accepted in
            precondition(Thread.isMainThread)
            guard !located else { return }; located = true
            guard let self else { lease.stop(); return }
            guard accepted, self.listening, !self.closed, self.generation == generation else {
                lease.stop(); self.lease = nil; self.busy = false; return
            }
            let release = self.release
            self.prepare(urls) { [weak self] result in
                precondition(Thread.isMainThread)
                // The SelectedFileStore now owns any successful URL access.
                lease.stop()
                guard let self else {
                    if case .success(let files) = result { release(files) }
                    return
                }
                self.lease = nil
                guard !self.closed, self.listening, self.generation == generation else {
                    if case .success(let files) = result { release(files) }
                    self.busy = false; return
                }
                switch result {
                case .failure:
                    self.busy = false; self.onError()
                case .success(let files):
                    self.pending = files
                    var decided = false
                    self.offer(files, x, y) { [weak self] accepted in
                        precondition(Thread.isMainThread)
                        guard !decided else { return }; decided = true
                        guard let self, !self.closed else { return }
                        // A true response owns the batch even if cancel crossed it.
                        self.finish(accepted: accepted)
                    }
                }
            }
        }
        return true
    }

    private func finish(accepted: Bool) {
        if !accepted { release(pending) }
        pending = []; busy = false
    }
    func close() {
        precondition(Thread.isMainThread)
        guard !closed else { return }
        cancel(); closed = true; finish(accepted: false)
    }

    private final class URLLease {
        private var stopAccess: (() -> Void)?
        init(_ stop: @escaping () -> Void) { stopAccess = stop }
        func stop() { let action = stopAccess; stopAccess = nil; action?() }
        deinit { stop() }
    }
}
